proposals/unified-show-diff.md

```diff
@@ -0,0 +1,134 @@
+# Title: Unified show/diff command for all output types
+
+- Enhancement Proposal PR: (leave this empty)
+- Contributors: dberenbaum
+
+# Summary
+
+Establish a unified command to combine and `show` or `diff` all outputs of
+specified types, including:
+
+* Metrics
+* Plots
+* Images (depends on https://github.com/iterative/dvc/discussions/5681)
+
+These would all be put under a single `dvc outputs show/diff` command.
+
+# Motivation
+
+* Enable users to `show` and `diff` multiple different output types within a
+  single command.
+* Reduce number of top-level commands by grouping these together.
+* Promote consistent UI, making it easier for users and for consumption by
+  downstream apps.
+* Generate flexible abstraction for future needs (additional output types or
+  additional functionality across all output types).
+
+# Detailed design
+
+### Definitions
+
+**Outputs**: include any `outs` specified in dvc stages, but only `metrics`,
+`plots`, and `images` will be supported for now. It's possible that other output
+types may be added in the future as needed.
+
+### Command Line Interface
+
+Combine all `show`/`diff` commands into one top-level command:
+
+```console
+usage: dvc outputs [--type] {show,diff}
+
+Commands to visualize and compare outputs (metrics, plots, and images).
+
+
+positional arguments:
+  COMMAND
+    show    Generate report from output files.
+    diff    Compare multiple versions of outputs in a single report.
+
+optional arguments:
+    --type  Restrict report to single output type (metrics, plots, images).
+```
+
```

Comment:
What would the expected output be for
```
show --type images
```

Are we just opening them in a browser like with plots?

Reply:
I also mentioned this in the other thread, but maybe we should consider defining an optional metadata/schema field for `mimetype`. If an output has a mimetype defined, we would then show it in a browser as needed.

The user would need to explicitly set `mimetype` themselves (on outputs they know will be images, or directories containing images for example). DVC would not attempt to do any auto detection (whether filename/extension based or by actually reading headers).

Reply:
I think plots and metrics would stay a special case, partially due to backwards-compatibility/existing behavior, and partially because they don't really fit into the mimetype idea anyways.

*Side note - if we go this route, we should avoid creating our own mimetypes - [mimetype](https://www.iana.org/assignments/media-types/media-types.xhtml) is a formal IANA standard. There would be some leeway to use something like `application/vnd.dvc.plots` but it's probably not worth it for us?*

Metrics are a special case for DVC since it's not just "yaml/json" files, it's "yaml/json" files containing only numeric values structured in a specific DVC-metrics compatible way.

plots (outputs) are a special case in the same way as metrics. And even though the actual generated plot is just a file with `mimetype: text/html`, we are really talking about the DVC `plots` (yaml/json) output and not the generated HTML file. The `plots` field also contains configuration information about things like the default template to use for a specific metric.

Reply:
I think it might be better to just leave binary/image support out of this proposal. There's already a separate discussion in https://github.com/iterative/dvc/discussions/5681, and it's a bit separate from the main point of the proposal.